matrix_sdk_crypto/src/key_request.rs

```diff
@@ -469,105 +508,189 @@ impl KeyRequestMachine {
     ///
     /// * `device` - The device that is requesting a session from us.
     ///
-    /// * `outbound_session` - If one still exists, the matching outbound
-    /// session that was used to create the inbound session that is being
-    /// requested.
-    fn should_share_session(
+    /// * `session` - The session that was requested to be shared.
+    async fn should_share_key(
         &self,
         device: &Device,
-        outbound_session: Option<&OutboundGroupSession>,
+        session: &InboundGroupSession,
     ) -> Result<Option<u32>, KeyshareDecision> {
-        if device.user_id() == self.user_id() {
+        let outbound_session = self
+            .outbound_group_sessions
+            .get_or_load(session.room_id())
+            .await
+            .ok()
+            .flatten()
+            .filter(|o| session.session_id() == o.session_id());
```

Comment:
Perhaps this dance could be factored out into a separate method which accepts both a room ID and the session ID?

Reply:
Sure, done.